modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/writer/ColumnValueWriter.java

```diff
@@ -29,15 +31,19 @@
 public class ColumnValueWriter {
 
 	public void write(
-			Object item, Map<String, Field> fieldMap,
+			Object item, Map<String, Field> fieldMap, List<String> fieldNames,
 			Consumer<Collection<?>> consumer)
 		throws IllegalAccessException {
 
 		Map<String, Object> columnNameValueMap = _getColumnNameValueMap(
-			item, fieldMap);
+			item, fieldMap, fieldNames);
 
 		if (!_firstLineWritten) {
-			consumer.accept(columnNameValueMap.keySet());
+			if (fieldNames.isEmpty()) {
+				fieldNames = new ArrayList<>(columnNameValueMap.keySet());
```

Comment:
@ivicac this does not change the old behavior at all.
In case fieldNames is empty, could simply because the user did not know he needs to supply it.
Then the fieldNames is exampled purely based on the first item. Then headers are fixed. If following items are having different numbers of locales or different types of locales, the exported file will be either corrupted by structure (value columns mismatch header columns) or corrupted by content (value does not match header).

I think this design is wrong, what you really want to set is what locales you want to export. A list of locales should be provided, if not provided only default to en_US (or any other locales you feel reasonable as default.) When extracting values, all locales must be tried, if missing just export as blank.